src/libraries/System.Text.Json/gen/Reflection/ReflectionExtensions.cs

```diff
@@ -45,5 +46,26 @@ private static bool HasJsonConstructorAttribute(ConstructorInfo constructorInfo)
 
             return false;
         }
+
+        public static Location? GetDiagnosticLocation(this Type type)
+        {
+            TypeWrapper? typeWrapper = type as TypeWrapper;
+            Debug.Assert(typeWrapper != null);
+            return typeWrapper.Location;
+        }
+
+        public static Location? GetDiagnosticLocation(this PropertyInfo propertyInfo)
+        {
+            PropertyInfoWrapper? propertyInfoWrapper = propertyInfo as PropertyInfoWrapper;
+            Debug.Assert(propertyInfoWrapper != null);
```

Comment:
If this invariant is always true, it begs the question why we need all the `System.Reflection` wrappers for Roslyn symbols. Not in scope for this PR, but in 7.0.0 I would recommend mapping roslyn symbols directly to sourcegen metadata to avoid abstraction leaks.

Reply:
In a few cases, we are able to share code between source-gen and reflection serializers since the symbols are expressed as reflection wrappers, e.g. the code in https://github.com/dotnet/runtime/blob/4cf86c2abfe9b2fad5ab2a48e4b0717ff0daf27a/src/libraries/System.Text.Json/Common/ReflectionExtensions.cs